### PR TITLE
Fix Kubernetes Agent resource serialization

### PR DESCRIPTION
**Pull request by flin-8**
[sc-71239]

The Kubernetes Agent resource couldn't actually be deserialized due to:
- No support for tiny types as the discriminator
- Incorrect resource type declaration

### CODE REVIEW

**Review by scme0 — APPROVED** [review]
Great find! I'm surprised this hasn't blown anything up in testing but I guess it's not really used on the Client side? 🤔

File: source/Octopus.Server.Client/Serialization/EndpointConverter.cs [hunk 1]
```diff
@@ -23,7 +23,7 @@ public class EndpointConverter : InheritedClassConverter<EndpointResource, Commu
                 {CommunicationStyle.Kubernetes, typeof (KubernetesEndpointResource)},
                 {CommunicationStyle.AzureServiceFabricCluster, typeof(ServiceFabricEndpointResource)},
                 {CommunicationStyle.StepPackage, typeof(StepPackageEndpointResource)},
-                {CommunicationStyle.KubernetesTentacle, typeof(KubernetesEndpointResource)}
+                {CommunicationStyle.KubernetesTentacle, typeof(KubernetesTentacleEndpointResource)}
```
Comment (flin-8) [inline source/Octopus.Server.Client/Serialization/EndpointConverter.cs:26]: This was declaring the wrong resource

File: source/Octopus.Server.Client/Serialization/InheritedClassConverter.cs [hunk 2]
```diff
@@ -73,12 +68,44 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
...
+        }
+
+        private static TDiscriminator GetDiscriminatingTypeFromStringValue(string derivedType)
+        {
+            if (typeof(TDiscriminator).IsEnum)
+            {
+                return (TDiscriminator)Enum.Parse(typeof(TDiscriminator), derivedType);
+            }
+
+            if (typeof(TDiscriminator).IsTinyType())
```
Comment (flin-8) [inline source/Octopus.Server.Client/Serialization/InheritedClassConverter.cs:90]: We didn't support tiny types before

File: source/Octopus.Server.Client/Serialization/InheritedClassConverter.cs [hunk 3]
```diff
@@ -73,12 +68,44 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
...
+            {
+                return (TDiscriminator)new TinyTypeConverter<TDiscriminator>().ConvertFrom(derivedType);
+            }
+
+            if (typeof(TDiscriminator) == typeof(string))
```
Comment (flin-8) [inline source/Octopus.Server.Client/Serialization/InheritedClassConverter.cs:95]: We don't strictly need this, no one has declared a string as the discriminator

File: source/Octopus.Server.Client/Serialization/InheritedClassConverter.cs [hunk 4]
```diff
@@ -73,12 +68,44 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
...
+
+            if (!DerivedTypeMappings.ContainsKey(discriminatingType))
+            {
+                throw new Exception($"Unable to determine type to deserialize. {TypeDesignatingPropertyName} `{discriminatingType}` does not map to a known type");
+            }
+
+            return DerivedTypeMappings[discriminatingType];
+        }
+
+        private static TDiscriminator GetDiscriminatingTypeFromStringValue(string derivedType)
```
Comment (flin-8) [inline source/Octopus.Server.Client/Serialization/InheritedClassConverter.cs:83]: I'm tossing up between writing unit tests for these 3 discriminator types (enum/tinytype/string) or have tests that ensure the real endpoints deserialize correctly

@APErebus @zentron

Comment (APErebus) [reply source/Octopus.Server.Client/Serialization/InheritedClassConverter.cs:83]: I'm tempted to say why not both (but I don't know how difficult it is to write tests for the endpoints)

**Review by APErebus — APPROVED** [review]
I'm happy with the current test coverage

**Comment by flin-8** [comment]
> Great find! I'm surprised this hasn't blown anything up in testing but I guess it's not really used on the Client side? 🤔

We don't deserialize on the client side using c#

File: source/Octopus.Server.Client/Serialization/InheritedClassConverter.cs (hunk 4 above)
Comment (flin-8) [reply source/Octopus.Server.Client/Serialization/InheritedClassConverter.cs:83]: I've added `InheritedClassConverterFixture` which tests the discriminator types

**Pull request merged**